File: chrome/renderer/extensions/context_menus_custom_bindings.cc
// Copyright (c) 2012 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/renderer/extensions/context_menus_custom_bindings.h"

#include "base/bind.h"
#include "chrome/common/extensions/extension_messages.h"
#include "content/public/renderer/render_thread.h"
#include "grit/renderer_resources.h"
#include "v8/include/v8.h"

namespace {

v8::Handle<v8::Value> GetNextContextMenuId(const v8::Arguments& args) {
  int context_menu_id = -1;
  content::RenderThread::Get()->Send(
      new ExtensionHostMsg_GenerateUniqueID(&context_menu_id));
  return v8::Integer::New(context_menu_id);
}

}  // namespace

namespace extensions {

ContextMenusCustomBindings::ContextMenusCustomBindings(
    Dispatcher* dispatcher, v8::Handle<v8::Context> v8_context)
    : ChromeV8Extension(dispatcher, v8_context) {
  RouteFunction("GetNextContextMenuId", base::Bind(&GetNextContextMenuId));
}

}  // extensions

File: extensions/renderer/resources/context_menus_custom_bindings.js
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// Custom binding for the contextMenus API.

var binding = require('binding').Binding.create('contextMenus');

var contextMenuNatives = requireNative('context_menus');
var sendRequest = require('sendRequest').sendRequest;
var Event = require('event_bindings').Event;

binding.registerCustomHook(function(bindingsAPI) {
  var apiFunctions = bindingsAPI.apiFunctions;

  var contextMenus = {};
  contextMenus.generatedIdHandlers = {};
  contextMenus.stringIdHandlers = {};
  var eventName = 'contextMenus';
  contextMenus.event = new Event(eventName);
  contextMenus.getIdFromCreateProperties = function(prop) {
    if (typeof(prop.id) !== 'undefined')
      return prop.id;
    return prop.generatedId;
  };
  contextMenus.handlersForId = function(id) {
    if (typeof(id) === 'number')
      return contextMenus.generatedIdHandlers;
    return contextMenus.stringIdHandlers;
  };
  contextMenus.ensureListenerSetup = function() {
    if (contextMenus.listening) {
      return;
    }
    contextMenus.listening = true;
    contextMenus.event.addListener(function() {
      // An extension context menu item has been clicked on - fire the onclick
      // if there is one.
      var id = arguments[0].menuItemId;
      var onclick = contextMenus.handlersForId(id)[id];
      if (onclick) {
        $Function.apply(onclick, null, arguments);
      }
    });
  };

  apiFunctions.setHandleRequest('create', function() {
    var args = arguments;
    var id = contextMenuNatives.GetNextContextMenuId();
    args[0].generatedId = id;
    var optArgs = {
      customCallback: this.customCallback,
    };
    sendRequest(this.name, args, this.definition.parameters, optArgs);
    return contextMenus.getIdFromCreateProperties(args[0]);
  });

  apiFunctions.setCustomCallback('create', function(name, request, response) {
    if (chrome.runtime.lastError) {
      return;
    }

    var id = contextMenus.getIdFromCreateProperties(request.args[0]);

    // Set up the onclick handler if we were passed one in the request.
    var onclick = request.args.length ? request.args[0].onclick : null;
    if (onclick) {
      contextMenus.ensureListenerSetup();
      contextMenus.handlersForId(id)[id] = onclick;
    }
  });

  apiFunctions.setCustomCallback('remove', function(name, request, response) {
    if (chrome.runtime.lastError) {
      return;
    }
    var id = request.args[0];
    delete contextMenus.handlersForId(id)[id];
  });

  apiFunctions.setCustomCallback('update', function(name, request, response) {
    if (chrome.runtime.lastError) {
      return;
    }
    var id = request.args[0];
    if (request.args[1].onclick) {
      contextMenus.handlersForId(id)[id] = request.args[1].onclick;
    }
  });

  apiFunctions.setCustomCallback('removeAll',
                                 function(name, request, response) {
    if (chrome.runtime.lastError) {
      return;
    }
    contextMenus.generatedIdHandlers = {};
    contextMenus.stringIdHandlers = {};
  });
});

exports.binding = binding.generate();
